api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,758 @@
+				s.Providers = append(s.Providers, newProvider)
+			}
+		}
+	}
+
+	if len(sorts[0].Providers) > len(sorts[1].Providers) {
+		return sorts[0].Providers, sorts[0].Errs
+	}
+
+	return sorts[1].Providers, sorts[1].Errs
+}
+
+func (e *EnvironmentScaler) getContainerResourceFromDeploy(deployID string) ([]ResourceConsumer, error) {
+	// use some kind of deploy cache
+	if consumers, ok := e.deployCache[deployID]; ok {
+		return consumers, nil
+	}
+
+	input := &ecs.DescribeTaskDefinitionInput{}
+	input.SetTaskDefinition(deployID)
+	output, err := e.Client.ECS.DescribeTaskDefinition(input)
+	if err != nil {
+		return nil, err
+	}
+
+	consumers := make([]ResourceConsumer, len(output.TaskDefinition.ContainerDefinitions))
+	for i, d := range output.TaskDefinition.ContainerDefinitions {
+		var memory bytesize.Bytesize
+
+		if aws.Int64Value(d.MemoryReservation) != 0 {
+			memory = bytesize.MiB * bytesize.Bytesize(aws.Int64Value(d.MemoryReservation))
+		}
+
+		if aws.Int64Value(d.Memory) != 0 {
+			memory = bytesize.MiB * bytesize.Bytesize(aws.Int64Value(d.Memory))
+		}
+
+		ports := []int{}
+		for _, p := range d.PortMappings {
+			if aws.Int64Value(p.HostPort) != 0 {
+				ports = append(ports, int(aws.Int64Value(p.HostPort)))
+			}
+		}
+
+		consumers[i] = ResourceConsumer{
+			ID:     "",
+			Memory: memory,
+			Ports:  ports,
+		}
+	}
+
+	e.deployCache[deployID] = consumers
+	return consumers, nil
+}
+
+func (e *EnvironmentScaler) getResourceConsumers_PendingServices(clusterName string) ([]ResourceConsumer, error) {
+	var resourceConsumers []ResourceConsumer
+
+	listServicesInput := &ecs.ListServicesInput{}
+	listServicesInput.SetCluster(clusterName)
+
+	serviceARNs := []*string{}
+	listServicesPagesFN := func(output *ecs.ListServicesOutput, lastPage bool) bool {
+		serviceARNs = append(serviceARNs, output.ServiceArns...)
+
+		return !lastPage
+	}
+
+	if err := e.Client.ECS.ListServicesPages(listServicesInput, listServicesPagesFN); err != nil {
+		return nil, err
+	}
+
+	services := []*ecs.Service{}
+	if len(serviceARNs) > 0 {
+		// The SDK states that you can specify up to 10 services in one DescribeServices operation:
+		// https://github.com/aws/aws-sdk-go/blob/ee1f179877b2daf2aaabf71fa900773bf8842253/service/ecs/api.go#L5420
+		// (aws-sdk-go version 1.12.19, as stated in layer0/Gopkg.toml)
+		for i := 0; i < len(serviceARNs); i += 10 {
+			end := i + 10
+
+			if end > len(serviceARNs) {
+				end = len(serviceARNs)
+			}
+
+			describeServicesInput := &ecs.DescribeServicesInput{}
+			describeServicesInput.SetCluster(clusterName)
+			describeServicesInput.SetServices(serviceARNs[i:end])
+
+			output, err := e.Client.ECS.DescribeServices(describeServicesInput)
+			if err != nil {
+				return nil, err
+			}
+
+			services = append(services, output.Services...)
+		}
+	}
+
+	for _, service := range services {
+		deployIDCopies := map[string]int64{}
+		for _, d := range service.Deployments {
+			desiredCount := aws.Int64Value(d.DesiredCount)
```

Comment:
I think it would be a good idea to move the old comment over as well some reading this for the first time would think, why don't we just use the pendingCount.

> // deployment.RunningCount is the number of containers already running on an instance
// deployment.PendingCount is the number of containers that are alraedy on an instance, but are being pulled
// we only care about containers that are not on instances yet

Reply:
✅